### PR TITLE
NEO-62182: Add uploadAemAsset functionality in acc js sdk

### PR DESCRIPTION
**Pull request by pk28yadav**
## Description
The function added provides functionality to call an POST API exposed by aemAssetDownload.jssp 
1. The method uploadAemAsset(assetDownloadUrl) is used to upload aem asset to campaign.
2. The assetDownloadUrl is the URL of the asset in AEM. This URL is used to download the assetInfo(signed asset link) from AEM and upload it to campaign.
3. The endpoint for this method is nms/aemAssetDownload.jssp
4. The method makes a post call to the above endpoint.
5. The method needs Authorization header to authenticate the call to campaign and also to download the asset from AEM.
6. The method returns the publishedURL of the asset uploaded on campaign publication server.
7. If publishing fails, an exception is thrown.

The called API is added in  https://git.corp.adobe.com/Campaign/ac/pull/15094 and
https://git.corp.adobe.com/Campaign/ac/pull/15341

## Related Issue

https://jira.corp.adobe.com/browse/NEO-62182

## Motivation and Context

It is required to add assets from asset essentials/ aem cloud service assets inside email designer(acrite).

## How Has This Been Tested?

Created a node application & called the above function by passing token & a correct assetDownloadUrl.
The code change doesn't seem to affect other areas of code.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] I have signed the [Adobe Open Source CLA](https://opensource.adobe.com/cla.html).
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by pk28yadav** [comment]
@mkiki Can you please add review for the PR?

**Review by mkiki — CHANGES_REQUESTED** [review]
Hi @pk28yadav, this looks good but can you add unit tests. We need 100% coverage with relevant assertions on the SDK. In this case it means you need to test with various mock answers (at least the successful path as well as common error scenarios such as AEM is down, asset does not exist, asset is too big, etc.) 

Also make sure this API is documented. Both us jsdoc and add a html/md file in the documentation. Documentation should provide examples, list error codes, etc. You'll find documentation here: https://opensource.adobe.com/acc-js-sdk/writeDoc.html

More generally, make sure all items in the commit checklist (https://opensource.adobe.com/acc-js-sdk/checkList.html) are validated

**Comment by pk28yadav** [comment]
@mkiki I have followed all the guidelines as in https://opensource.adobe.com/acc-js-sdk/checkList.html . Please add review.
I will try to add installation steps for ruby & doc setup for linux in another PR as they were a bit different from macos ones & took me some time.

**Comment by pk28yadav** [comment]
@mkiki gentle reminder for adding review, thanks.

File: docs/uploadAemAsset.html [hunk 1]
```diff
@@ -0,0 +1,55 @@
+---
+layout: page
+title: AEM asset upload API
+---
+
+<p>Campaign API for downloading & uploading(publishing) AEM cloud asset on campaign publication server. Just like all APIs in the SDK, it's been wrapped into a function and will return a JSON object containing publishedUrl.</p>
+<p class="warning">Note: This API is restricted to be used from acc-web-ui only. It is used when an asset is selected from
```
Comment (mkiki) [inline docs/uploadAemAsset.html:7]: I do not like this very much? The SDK is supposed to be a general purpose SDK and should not depend on ACC Web UI or any particular usage. Why is there this restriction?

File: docs/uploadAemAsset.html [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+console.log(`${version.description} version ${version.version}`);
+
+<span class="comment">// Logon to a Campaign instance with ImsBearerToken</span>
+const connectionParameters = sdk.ConnectionParameters.ofImsBearerToken("http://acc-sdk:8080", "ey...", options);
+console.log(connectionParameters);
+
+const client = await sdk.init(connectionParameters);
+await client.logon();
+
+<span class="comment">// This url is download url link provided by acrite from selection handlebar</span>
```
Comment (mkiki) [inline docs/uploadAemAsset.html:25]: Remove reference to "Acrite". Acrite is an internal code name. SDK is Open Source and should not depend on internal projects.

File: docs/uploadAemAsset.html [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+<span class="comment">// when asset is selected in asset-picker UI component.</span>
+const url = "https://abcd.adobeaemcloud/asset/xyz.png";
+
+try {
+  const response = await client.fileUploader.uploadAemAsset(url);
+  console.log(response);
+}catch(e){
```
Comment (mkiki) [inline docs/uploadAemAsset.html:32]: I don't think this try/catch is useful. For the sake of the simplicity of the example, could you remove it?

File: docs/uploadAemAsset.html [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+    console.log(e);
+}
+</pre>
+
+<p>will return in this format</p>
+<pre class="code">
+{
+    "publishedURL": "http://trk-inst-panyadav.corp.adobe.com/res/trk-inst/12f6710969d74ad965a333d31cf8c06d.jpeg"
```
Comment (mkiki) [inline docs/uploadAemAsset.html:40]: I would not put any real URL or internal Adobe DNS here. Could you use a dummy name, for instance http://trk-my-instance.my.domain.com/res/...

File: docs/uploadAemAsset.html [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+}
+</pre>
+
+<p class="warning">Note: This API can only be authenticated using ImsBearerToken because the same Ims user token
```
Comment (mkiki) [inline docs/uploadAemAsset.html:44]: Great call out.

File: src/client.js [hunk 6]
```diff
@@ -2002,7 +2037,7 @@ class Client {
             if (this._traceAPICalls)
                 console.log("HTTP//failure", err);
             this._notifyObservers((observer) => observer.onHTTPCallFailure && observer.onHTTPCallFailure(request, err) );
-            const ex = makeCampaignException({ request:request, reqponse:err.response }, err);
+            const ex = makeCampaignException({ request:request, response:err.response }, err);
```
Comment (mkiki) [inline src/client.js:2040]: Nice catch, thank you

File: src/campaign.js [hunk 7]
```diff
@@ -226,13 +227,14 @@ function makeCampaignException(call, err) {
 
   if (err.statusCode && ctor && ctor.name == "HttpError") {
     var faultString = err.statusText;
-    var details = err.data;
+    var details = typeof err.data == 'object' ? JSON.stringify(err.data) : err.data;
```
Comment (mkiki) [inline src/campaign.js:230]: Why is this needed? I do not see anywhere where an object would be returned in err.data?

**Review by mkiki — CHANGES_REQUESTED** [review]

File: docs/uploadAemAsset.html (hunk 1 above)
Comment (pk28yadav) [reply docs/uploadAemAsset.html:7]: Hi, The API isn't restricting us to be only called from web-ui, but i can't find any usecase of it other than web-ui.
Code added is for feature which shows aem cloud or asset essentials asset library within email-designer acrite.
This API gets an assetInfo url instead of direct download url from acrite(acrite gets it from https://git.corp.adobe.com/CQ/assets-selectors)- assetInfo url when called with GET method & auth token returns a short lived signed url for downloading the asset.

File: docs/uploadAemAsset.html (hunk 2 above)
Comment (pk28yadav) [reply docs/uploadAemAsset.html:25]: I will remove the reference.

File: src/campaign.js (hunk 7 above)
Comment (pk28yadav) [reply src/campaign.js:230]: it is object when you explicitly return json output as response from jssp API.

**Comment by pk28yadav** [comment]
@mkiki Thanks for review, added changes

**Review by mkiki — APPROVED** [review]
